### PR TITLE
REF: stricter types for RangeIndex._simple_new

### PR DESCRIPTION
**Pull request by jbrockmendel**
xref #31055.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3105,7 +3106,10 @@ def reindex(self, target, method=None, level=None, limit=None, tolerance=None):
         if not isinstance(target, Index) and len(target) == 0:
             attrs = self._get_attributes_dict()
             attrs.pop("freq", None)  # don't preserve freq
-            values = self._data[:0]  # appropriately-dtyped empty array
+            if isinstance(self, ABCRangeIndex):
```
Comment (WillAyd) [inline pandas/core/indexes/base.py]: What does this need to change for?

**Review by WillAyd — COMMENTED** [review]
function cleanup looks good; question on the reindex

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: `self._data[:0]` is an ndarray for RangeIndex, so we can't pass it to self._simple_new below

**Comment by jreback** [comment]
sure

**Pull request merged**